ocaml/typing/ctype.ml

```diff
@@ -1588,6 +1588,17 @@ let with_locality locality m =
   Alloc.submode_exn (Alloc.meet_with (Comonadic Areality) Locality.Const.min m) m';
   m'
 
+let curry_mode alloc arg : Alloc.Const.t =
+  let acc =
+    Alloc.Const.join
+      (Alloc.Const.close_over arg)
+      (Alloc.Const.partial_apply alloc)
+  in
+  (* Arrow types cross uniqueness axis. Therefore, when user writes an
+  A -> B -> C (to be used as constraint on something), we should make
+  (B -> C) shared. A proper way to do this is via modal kinds. *)
```

Comment:
I would expect we'd want to make `B -> C` unique, not shared.

Reply:
In fact neither is correct. We don't know in which variance this `B->C` will be, so we don't know which direction to push.

For example: `let f : A -> B -> C = ...`, in this case we want to push to `unique` for a stronger type.
Another example: `let f (g : A -> B -> C) = ...` in this case we want to push to `shared`, so that `f` is more permissive in its argument.

The current approach of making it `shared` just happens to work, because `shared` is legacy and we haven't used the axis much.

To properly support this, we need mode crossing on arrow types.

Reply:
I think what you're saying could maybe work? But probably not, because somewhere we'll have a `'a` that gets discovered to be `t1 -> t2` but it won't be unique.

But actually I think we do have subtyping for function arguments:

```ocaml
let f (g : string -> local_ (string -> string)) = g "hi" "bye"

let h s1 s2 = "foo"

let x = f h
```

This works, because global < local. So I think shared is fine here. But this has nothing to do with mode-crossing. The comment here is about mode-crossing, so that's maybe where I'm confused.

Reply:
Huh indeed. I had the wrong impression.

Now consider the example:
```
let g : (A -> B -> C) = ...
```
if we say `B -> C` is `shared`, we will be producing the `g` weaker than it needs to be. And now if we have:
```
let f (g : A -> unique_ (B -> C)) = ...
```
`f g` will not work, because mode crossing doesn't work deeply into arrows. But I suspect this won't cause much trouble in practice, because the author of `f` shouldn't ask `B -> C` to be `unique_`. Instead, they should leave it as default which is `shared`, and mode crossing it to `unique` at the location where `B -> C` is a real value (instead of the return of a function).

Reply:
Yes I agree with all that. We could perhaps add mode-crossing considerations to work in functions (I think this would fit into the current approach to arrow subtyping), but we probably won't ever have to.

Reply:
I still don't quite see what this conversation has to do with the line of code / comment highlighted, though...

Reply:
Yes indeed. The new commit has rewritten this comment.